apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/SamplingProfiler.java

```diff
@@ -0,0 +1,657 @@
+            }
+        }
+    }
+
+    private void resetActivationEventBuffer() {
+        ((Buffer) activationEventBuffer).clear();
+    }
+
+    EventPoller.PollState consumeActivationEventsFromRingBufferAndWriteToFile() throws Exception {
+        return poller.poll(writeActivationEventToFileHandler);
+    }
+
+    private void processTraces(File file) throws IOException {
+        if (activationEventBuffer.position() == 0) {
+            logger.debug("No activation events during this period. Skip processing stack traces.");
+            return;
+        }
+        long start = System.nanoTime();
+        List<WildcardMatcher> excludedClasses = config.getExcludedClasses();
+        List<WildcardMatcher> includedClasses = config.getIncludedClasses();
+        try {
+            jfrParser.parse(file, excludedClasses, includedClasses);
+            startProcessingActivationEventsFile();
+            final SortedSet<StackTraceEvent> stackTraceEvents = getStackTraceEvents(jfrParser);
+            if (logger.isDebugEnabled()) {
+                logger.debug("Processing {} stack traces", stackTraceEvents.size());
+            }
+            List<StackFrame> stackFrames = new ArrayList<>();
+            ElasticApmTracer tracer = this.tracer;
+            ActivationEvent event = new ActivationEvent();
+            for (StackTraceEvent stackTrace : stackTraceEvents) {
+                processActivationEventsUpTo(stackTrace.nanoTime, event);
+                CallTree.Root root = profiledThreads.get(stackTrace.threadId);
+                if (root != null) {
+                    jfrParser.resolveStackTrace(stackTrace.stackTraceId, true, stackFrames, MAX_STACK_DEPTH);
+                    if (stackFrames.size() == MAX_STACK_DEPTH) {
+                        logger.debug("Max stack depth reached. Set profiling_included_classes or profiling_excluded_classes.");
+                    }
+                    root.addStackTrace(tracer, stackFrames, stackTrace.nanoTime);
+                }
+                stackFrames.clear();
+            }
+        } finally {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Processing traces took {}µs", (System.nanoTime() - start) / 1000);
+            }
+            jfrParser.resetState();
+        }
+    }
+
+    /**
+     * Returns stack trace events of relevant threads sorted by timestamp.
+     * The events in the JFR file are not in order.
+     * Even for the same thread, a more recent event might come before an older event.
+     * In order to be able to correlate stack trace events and activation events, both need to be in order.
+     *
+     * Returns only events for threads where at least one activation happened
+     */
+    private SortedSet<StackTraceEvent> getStackTraceEvents(JfrParser jfrParser) throws IOException {
+        final LongHashSet nativeThreadIds = threadMapper.getNativeThreadIds();
+        final SortedSet<StackTraceEvent> stackTraceEvents = new TreeSet<>();
```

Comment:
IMO, a regular `List`, sorted afterwards, will be more space-efficient, and will preserve StackTraceEvents with duplicate timestamps (although they are unlikely).